pkg/common/utils.go

```diff
@@ -115,6 +123,56 @@ func ConvertSITimeout(millis int64) time.Duration {
 	return time.Duration(result)
 }
 
+// ConvertSITimeoutWithAdjustment Similar to ConvertSITimeout, but this function also adjusts the timeout if
+// "creationTime" is defined. It's used during Yunikorn restart, in order to properly track how long a placeholder pod should
+// be in "Running" state.
+func ConvertSITimeoutWithAdjustment(siApp *si.AddApplicationRequest) time.Duration {
+	result := ConvertSITimeout(siApp.ExecutionTimeoutMilliSeconds)
+	adjusted := adjustTimeout(result, siApp)
+
+	return adjusted
+}
+
+func adjustTimeout(timeout time.Duration, siApp *si.AddApplicationRequest) time.Duration {
+	creationTimeTag := siApp.Tags[interfaceCommon.DomainYuniKorn+CreationTime]
+	if creationTimeTag == "" {
+		return timeout
+	}
+
+	created := ConvertSITimestamp(creationTimeTag)
+	if created == Undefined {
+		return timeout
+	}
+	now := currentTime()
+	expectedTimeout := created.Add(timeout)
+
+	if now.After(expectedTimeout) {
+		log.Logger().Info("Placeholder timeout reached - expected timeout is in the past",
+			zap.Duration("timeout duration", timeout),
+			zap.Time("creation time", created),
+			zap.Time("expected timeout", expectedTimeout))
+		return time.Millisecond // smallest allowed timeout value
+	}
+
+	log.Logger().Info("Adjusting placeholder timeout",
+		zap.Duration("timeout duration", timeout),
+		zap.Time("creation time", created),
+		zap.Time("expected timeout", expectedTimeout))
+
+	return expectedTimeout.Sub(now)
+}
+
+func ConvertSITimestamp(ts string) time.Time {
```

Comment:
Design decision to be made: do we support negative numbers in the timestamp? i.e. dates before `1970-01-01 00:00:00`
Also not sure if you want to print a warning when the string passed in is empty. Pollutes the logs and would happen often.

Reply:
Fixed